### PR TITLE
[TT-7902] Detailed Tracing

### PR DESCRIPTION
**Pull request by mativm02**
## Description

Adding `detailed_tracing` field to API Definition. If this is set to true, then each OpenTelemetry trace will propagate the context and multiple spans will be shown. Otherwise, only the main span and a span showing the request to the upstream service will be present.
## Related Issue
https://tyktech.atlassian.net/browse/TT-7902?atlOrigin=eyJpIjoiMjgwMDY3YTQ5Nzk0NDAyODlkODUxOTUzMzc5NTZhZDgiLCJwIjoiaiJ9

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Refactoring or add test (improvements in base code or adds test coverage to functionality)

## Checklist

- [ ] I ensured that the documentation is up to date
- [ ] I explained why this PR updates go.mod in detail with reasoning why it's required
- [ ] I would like a code coverage CI quality gate exception and have explained why

### CODE REVIEW

**Comment by Tyk-ITS** [comment]
**API tests result: failure** :no_entry_sign: 
Branch used: refs/pull/5294/merge
Commit:  
Triggered by: pull_request (@mativm02)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/5487178244)

**Comment by Tyk-ITS** [comment]
**API tests result: failure** :no_entry_sign: 
Branch used: refs/pull/5294/merge
Commit: 10c54c7086299040c579a05e0913b10658f86559 
Triggered by: pull_request (@mativm02)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/5487512535)

**Comment by Tyk-ITS** [comment]
**API tests result: failure** :no_entry_sign: 
Branch used: refs/pull/5294/merge
Commit: 45f08e1f5241309e07ce5a591e056de416053737 
Triggered by: pull_request (@mativm02)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/5487748354)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/5294/merge
Commit: 0b50bb79639e658c9f208dc48abd748d9131681b 
Triggered by: pull_request (@mativm02)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/5508224877)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/5294/merge
Commit: 6027a77f3c008aad12f172b09c64c2cfd8adbe9e 
Triggered by: pull_request (@mativm02)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/5508221362)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/5294/merge
Commit: b828b57dca6a1900af7000328bdd9c729733f28d 
Triggered by: pull_request (@mativm02)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/5508302882)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/5294/merge
Commit: 6bd64b74757361f81f55c158a639a1220a147250 
Triggered by: pull_request (@mativm02)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/5519565037)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/5294/merge
Commit: f0d6d7c29845b317b67e194ce20141064c7f76aa 
Triggered by: pull_request (@mativm02)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/5519566347)

**Comment by Tyk-ITS** [comment]
**API tests result: failure** :no_entry_sign: 
Branch used: refs/pull/5294/merge
Commit: eaa1139723147eb6e66aa9d14880552a01f0612a 
Triggered by: pull_request (@mativm02)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/5519432494)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/5294/merge
Commit: b24ad00a9c54df6b1b50a4c002a4749e73c0e24d 
Triggered by: pull_request (@mativm02)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/5519671645)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/5294/merge
Commit: b24ad00a9c54df6b1b50a4c002a4749e73c0e24d 
Triggered by: pull_request (@furkansenharputlu)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/5519671630)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/5294/merge
Commit: ae25d2adab9fea9799ca8e91235871599d008d8f 
Triggered by: pull_request (@mativm02)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/5520229619)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/5294/merge
Commit: 81d870224d3d1f52b9122f6a1899ed5341a123eb 
Triggered by: pull_request (@mativm02)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/5520338853)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/5294/merge
Commit: de7556f903d6c4c6a7d7e95dd6557c8a0a0770f3 
Triggered by: pull_request (@mativm02)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/5531411752)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/5294/merge
Commit: bb582fc45c0c401c866c8151e26216f0c30e72d2 
Triggered by: pull_request (@mativm02)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/5610860950)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/5294/merge
Commit: 5ffd2c65949485f184826a2ddbcf5cacc1c4dbba 
Triggered by: pull_request (@mativm02)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/5612589518)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/5294/merge
Commit: 1ee9a1f671b28ae0980354778289051952642b4d 
Triggered by: pull_request (@mativm02)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/5622271274)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/5294/merge
Commit: 3ff2a8b1a827616d438567ae27550f7d2e573421 
Triggered by: pull_request (@mativm02)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/5622289665)

File: Taskfile.yml [hunk 1]
```diff
@@ -13,8 +13,7 @@ tasks:
     sources:
       - go.mod
       - go.sum
-      - './**/*.go'
-
+      - "./**/*.go"
```
Comment (mativm02) [inline Taskfile.yml:16]: change made by the linter

File: Taskfile.yml [hunk 2]
```diff
@@ -63,16 +61,21 @@ tasks:
...
     cmds:
-      - task: test:opentelemetry:setup
-      - task: test:opentelemetry:test
-      - task: test:opentelemetry:teardown
\ No newline at end of file
+      - |
+        set -e
+        task test:opentelemetry:setup
+        trap 'task test:opentelemetry:teardown' EXIT
+        task test:opentelemetry:test
```
Comment (mativm02) [inline Taskfile.yml:81]: I'm ensuring teardown is executed no matter if the test failed or not

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/5294/merge
Commit: c3284a5e1b2e5c713f79435f38b78c1b3aed5188 
Triggered by: pull_request (@mativm02)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/5645053376)

File: gateway/middleware.go [hunk 3]
```diff
@@ -66,17 +69,21 @@ func (tr TraceMiddleware) ProcessRequest(w http.ResponseWriter, r *http.Request,
...
+		cfg := baseMw.Gw.GetConfig()
+		if cfg.OpenTelemetry.Enabled {
+			var span tyktrace.Span
+			if baseMw.Spec.DetailedTracing {
+				var ctx context.Context
+				ctx, span = baseMw.Gw.TracerProvider.Tracer().Start(r.Context(), tr.Name())
+				defer span.End()
+				setContext(r, ctx)
+			} else {
+				span = tyktrace.SpanFromContext(r.Context())
```
Comment (tbuchaillot) [inline gateway/middleware.go:81]: Move this function to otel internal pkg, so we don't have to import /opentelemetry/trace here

File: apidef/schema.go [hunk 4]
```diff
@@ -724,7 +724,10 @@ const Schema = `{
         },
 		"is_oas": {
 			"type": "boolean"
-		}
+		},
+        "detailed_tracing": {
```
Comment (tbuchaillot) [inline apidef/schema.go:728]: Is this linted?

**Review by tbuchaillot — CHANGES_REQUESTED** [review]
Please add a test case  here with detailed spans  https://github.com/TykTechnologies/tyk/blob/master/gateway/api_loader_test.go#L434

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/5294/merge
Commit: 6b8465653b8e0bb75b073222697205569d724d0f 
Triggered by: pull_request (@mativm02)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/5646100321)

File: apidef/schema.go (hunk 4 above)
Comment (mativm02) [reply apidef/schema.go:728]: it is!

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/5294/merge
Commit: 6de24eeeec280883ca4a5a93d2b101cbe0f6ff9f 
Triggered by: pull_request (@mativm02)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/5657988450)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/5294/merge
Commit: 6f5619b20e5cb7a4fbb90afe9175ae59e7eec7fd 
Triggered by: pull_request (@mativm02)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/5658068202)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/5294/merge
Commit: 6b8465653b8e0bb75b073222697205569d724d0f 
Triggered by: pull_request (@mativm02)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/5715836627)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/5294/merge
Commit: 8c83546eca3609b0cdcddcfbc64e2d1b1924a3ab 
Triggered by: pull_request (@mativm02)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/5715847807)

**Pull request merged**